source/lib/src/gpu/prod_force.cu

```diff
@@ -43,92 +43,146 @@ __global__ void force_deriv_wrt_center_atom(FPTYPE* force,
-  const int ndescrpt = nnei * 4;
-  if (idy >= nnei) {
-    return;
-  }
-  // deriv wrt neighbors
-  int j_idx = nlist[idx * nnei + idy];
-  if (j_idx < 0) {
+template <typename FPTYPE,
+          bool radial_only_ = false,
+          int shared_memory_block_ = 32>
+__global__ void force_deriv_wrt_neighbors(FPTYPE* force,
+                                          const FPTYPE* net_deriv,
+                                          const FPTYPE* in_deriv,
+                                          const int* nlist,
+                                          const int nframes,
+                                          const int nloc,
+                                          const int nall,
+                                          const int nnei) {
+  // limited to 2 billions atoms and 2 billions frames
+  const int atom_id = blockIdx.x;
+  const int frame_id = blockIdx.z * gridDim.y + blockIdx.y;
+
+  if (frame_id >= nframes) {
     return;
   }
-  FPTYPE force_tmp = (FPTYPE)0.;
-  for (int idw = 0; idw < 4; ++idw) {
-    force_tmp += net_deriv[idx * ndescrpt + idy * 4 + idw] *
-                 in_deriv[idx * ndescrpt * 3 + (idy * 4 + idw) * 3 + idz];
+
+  const int ndescrpt = nnei * ((radial_only_) ? (1) : (4));
+
+  // define various pointers for a specific frame.
+  const FPTYPE* frame_net_deriv_ = &net_deriv[frame_id * nloc * ndescrpt];
+  const FPTYPE* frame_in_deriv_ = &in_deriv[frame_id * nloc * ndescrpt * 3];
+  const int* frame_neighbor_list_ = &nlist[frame_id * nnei * nloc];
+  FPTYPE force_tmp[3] = {(FPTYPE)0., (FPTYPE)0., (FPTYPE)0.};
+
+  for (int neighbor_id = threadIdx.x; neighbor_id < nnei;
+       neighbor_id += blockDim.x) {
+    // collect all terms $\partial E_j / \partial D_{ji} \nabla_R_j D_{ji}$
+    // where the atom i is a neighbor of the atom j.
+    //
+    // Go through all neighbors of atom i, locate the position of
+    // the atom i in the neighbor list of the atom j and retrieve all necessary
+    // information.
+
+    const int atom_j = frame_neighbor_list_[atom_id * nnei + neighbor_id];
+
+    // The neighbors of a given atom are sorted by type and each resulting list
+    // is separated from the other by a series of -1. More details about the
+    // sorting can be found in https://doi.org/10.1016/j.cpc.2020.107624
+    //
+    // To illustrate this, take the neigbhors of a given atom of type a (in a
+    // system with two atoms type a and b) deepmd stores the neighbors as
+    //
+    // [neighbors list of type a], -1, -1, -1, ...., [neighbor list of type b],
+    // -1, -1, -1, .....
+
+    if (atom_j < 0) {
+      continue;
+    }
+
+    const int* nei_nei_list_ = &frame_neighbor_list_[atom_j * nnei];
+    int atom_id_position = 0;
+
+    // search the index of the atom i in the local neighbor list of atom j
+    for (atom_id_position = 0; atom_id_position < nnei; atom_id_position++) {
+      if (nei_nei_list_[atom_id_position] == atom_id) {
+        break;
+      }
+    }
```

Comment:
The complexity of the index searching is of order N_nei, which does not present in the atomic operation implementation. Does it have an observable side effect on the performance of the `prod_force` operator?

Reply:
I could not observe any side effect except fluctuations of 5 % in performance. The gain is that these operators are deterministic by default which is worth the 5% potential penalty (or less) introduced by this code change.